pydatastructs/strings/suffix_tree.py

```diff
@@ -0,0 +1,144 @@
+from pydatastructs.utils.misc_util import (
+    SuffixTreeNode, SuffixTreeEdge, Suffix, Backend, raise_if_backend_is_not_python)
+
+__all__ = [
+    'SuffixTree'
+]
+
+
+# Ukkonen's algorithm gives a O(n) + O(k) contruction time for a suffix tree,
+# where n is the length of the string and k is the size of the alphabet of that string.
+# Ukkonen's is an online algorithm,
+# processing the input sequentially and producing a valid suffix tree at each character.
+
+
+class SuffixTree(object):
+    """A suffix tree for string matching. Uses Ukkonen's algorithm
+    for construction.
+    """
+    @classmethod
+    def methods(cls):
+        return ['__new__', '__init__', '__repr__',
+                'find_substring', 'has_substring']
+
+    def __new__ (cls, *args, **kwargs):
+        instance = super().__new__(cls)
+        return instance
+
+    def __init__(self, string, case_insensitive=False):
+
+        self.string = string
+        self.case_insensitive = case_insensitive
+        self.N = len(string) - 1
+        self.nodes = [SuffixTreeNode()]
+        self.edges = {}
+        self.active = Suffix(0, 0, -1)
+        if self.case_insensitive:
+            self.string = self.string.lower()
+        for i in range(len(string)):
+            self._add_prefix(i)
```

Comment:
You can comine these too. You can do `instance.N`, `instance.nodes`, etc in `__new__` and then at the end do `return instance`.

Reply:
See the body of `__new__` below for example,

https://github.com/codezonediitj/pydatastructs/blob/2482adb276055403485266d6d02a1489451c8e76/pydatastructs/trees/binary_trees.py#L59-L74